### PR TITLE
[vividus] Add ability to use variables in templates

### PR DESCRIPTION
**Pull request by uarlouski**
Relates #1545

### CODE REVIEW

**Review by valfirst — CHANGES_REQUESTED** [review]
Please add docs

File: vividus/src/main/java/org/vividus/bdd/steps/BddVariableSteps.java [hunk 1]
```diff
@@ -122,7 +122,9 @@ protected boolean compareListsOfMaps(Object variable1, Object variable2)
     public void initVariableUsingTemplate(Set<VariableScope> scopes, String variableName, String templatePath,
             ExamplesTable templateParameters) throws IOException, TemplateException
     {
-        Map<String, List<String>> dataModel = MapUtils.convertExamplesTableToMap(templateParameters);
+        Map<String, Object> dataModel = bddVariableContext.getVariables();
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/bdd/steps/BddVariableSteps.java:125]: please update javadocs according to the changes

**Review by valfirst — CHANGES_REQUESTED** [review]

File: docs/modules/commons/pages/vividus-steps.adoc [hunk 2]
```diff
@@ -92,19 +86,50 @@ Waits during specified period for debug purposes
...
 ----
-Scenario: Wait for debug
 When I wait `PT30S` for debug
 ----
+
+=== Initialize variable using template
+
+Initializes a variable with a result of the processed https://freemarker.apache.org/docs/dgui_template_overallstructure.html[Freemarker template]
+
+It's allowed to use `global`, `scenario`, `story` and other variable types within templates by referring them using the variable reference notation. Note than the parameters passed to the step take precedence over variables.
```
Comment (valfirst) [inline docs/modules/commons/pages/vividus-steps.adoc:106]:
```
...to use `global`, `scenario`, `story` and other variable types...
```
->
```
...to use `global`, `next batches`, `scenario` and `story` variables...
```

Comment (valfirst) [inline docs/modules/commons/pages/vividus-steps.adoc:106]: `than` -> `that`

File: docs/modules/commons/pages/vividus-steps.adoc [hunk 3]
```diff
@@ -92,19 +86,50 @@ Waits during specified period for debug purposes
...
+
+[source,gherkin]
+----
+Given I initialize the $scopes variable `$variableName` using template `$templatePath` with parameters:$templateParameters
+----
+
+* `$scopes` - xref:parameters:variable-scope.adoc[The comma-separated set of the variables scopes].
+* `$variableName` - The variable name to store a result
+* `$templatePath` - The relative to `src/main/resources` path to a template
+* `$templateParameters` - The template parameters
```
Comment (valfirst) [inline docs/modules/commons/pages/vividus-steps.adoc:116]: add dots `.` to the end of all parameters descriptions

File: docs/modules/commons/pages/vividus-steps.adoc [hunk 4]
```diff
@@ -92,19 +86,50 @@ Waits during specified period for debug purposes
...
+  "name": "${name[0]}",
+  "race": "${race[0]}",
+  "age": ${age}
+}
+----
+
+.Generate JSON from the personal info template
+[source,gherkin]
+----
+When I initialize the $scopes variable `age` with value `4510`
```
Comment (valfirst) [inline docs/modules/commons/pages/vividus-steps.adoc:131]: ` $scopes` -> `scenario`

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus/src/main/java/org/vividus/bdd/steps/BddVariableSteps.java [hunk 5]
```diff
@@ -83,6 +83,9 @@ protected boolean compareListsOfMaps(Object variable1, Object variable2)
 
     /**
      * This step initializes BDD variable with a result of given template processing
+     * It's allowed to use <i>global</i>, <i>scenario</i>, <i>story</i> and other variable types within templates
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/bdd/steps/BddVariableSteps.java:86]: https://github.com/vividus-framework/vividus/pull/1550#discussion_r610850507

File: vividus/src/main/java/org/vividus/bdd/steps/BddVariableSteps.java [hunk 6]
```diff
@@ -83,6 +83,9 @@ protected boolean compareListsOfMaps(Object variable1, Object variable2)
...
+     * by referring them using the variable reference notation. Note than the parameters passed to the step take
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/bdd/steps/BddVariableSteps.java:87]: https://github.com/vividus-framework/vividus/pull/1550/files#r610850715

File: vividus/src/main/java/org/vividus/bdd/steps/BddVariableSteps.java [hunk 7]
```diff
@@ -83,6 +83,10 @@ protected boolean compareListsOfMaps(Object variable1, Object variable2)
 
     /**
      * This step initializes BDD variable with a result of given template processing
+     * It's allowed to use <i>global</i>, <i>next batches</i>, <i>scenario</i> and <i>story</i> variables within
+     * templates by referring them using the variable reference notation. Note that the parameters passed to the
+     * step take
+     * precedence over the variables.
```
Comment (valfirst) [inline vividus/src/main/java/org/vividus/bdd/steps/BddVariableSteps.java:89]: it can be moved to the previous line

**Review by valfirst — APPROVED** [review]

**Review by ikalinin1 — APPROVED** [review]

**Pull request merged**